Update design section in mdbook

bobbinth opened the pull request:
This PR updates the Design section in mdBook documentation. Specifically:

- Adds the overall description of the VM design (`/design/main.md`).
- Adds a section on operation constraints and operation flags.
- Updates description of stack design section.
- Cleans up descriptions of all stack operation constraints.

After this PR the design section in mdbook docs is nearly complete. The outstanding items are:
- Rewrite chiplet intro section (`/design/chiplets/main.md`).
- Write multiset checks section (`/design/multiset.md`) - this section is currently just stabbed out.

Once merged, this PR will address #190 and #203

grjte commented on docs/src/design/stack/crypto_ops.md:
```diff
@@ -0,0 +1,122 @@
+v_{input} = \alpha_0 + \alpha_1 \cdot op_{mpver} + \alpha_2 \cdot r + \alpha_3 \cdot i + \sum_{j=0}^3 \alpha_{j+8} \cdot s_{6 - j}
+$$
+
+$$
+v_{output} = \alpha_0 + \alpha_1 \cdot op_{rethash} + \alpha_2 \cdot (h_0 + 8 \cdot s_0 - 1) + \sum_{j=0}^3\alpha_{j + 8} \cdot s_{10 - j}
+$$
```
Shouldn't the stack index here be from s_9 down to s_6?

bobbinth replied:
Good catch! Fixed!

grjte commented on docs/src/design/stack/crypto_ops.md:
```diff
@@ -0,0 +1,122 @@
+v_{outputnew} = \alpha_0 + \alpha_1 \cdot op_{rethash} + \alpha_2 \cdot (r + 2 \cdot 8 \cdot d - 1) + \sum_{j=0}^3\alpha_{j + 8} \cdot s_{9 - j}'
+$$
+
+In the above, the first two expressions correspond to inputs and outputs for verifying the Merkle path between the old node value and the old tree root, while the last two expressions correspond to inputs and outputs for verifying the Merkle path between the new node value and the new tree root. Hash chiplet ensures the same set of sibling nodes are uses in both of these computations.
```
nit: "...The hash chiplet ensures the same set of sibling nodes are used in both of these computations."

bobbinth replied:
Fixed!

grjte commented on docs/src/design/stack/crypto_ops.md:
```diff
@@ -0,0 +1,122 @@
+v_{output} = \alpha_0 + \alpha_1 \cdot op_{retstate} + \alpha_2 \cdot (h_0 + 7) + \sum_{j=0}^{11} (\alpha_{i+4} \cdot s_{11-i}')
+$$
+
+In the above, $op_{linhash}$ and $op_{retstate}$ are the unique [operation label](../chiplets/main.md#operation-labels) for initiating a linear hash and reading the full state of the hasher respectively. Also note that the term for $\alpha_3$ is missing from the above expressions because for Rescue Prime permutation computation the index column is expected to be set to $0$.
```
nit: "... are the unique operation labels..." (this comes up a few times, so if you do want to change it then maybe just search & replace. I won't mark the others)

bobbinth replied:
Fixed!

grjte commented on docs/src/design/stack/crypto_ops.md:
```diff
@@ -0,0 +1,122 @@
+b_{chip}' \cdot v_{inputold} \cdot v_{outputold} \cdot v_{inputnew} \cdot v_{outputnew} = b_{chip} \text{ | degree } = 5
+$$
+
+The above constrain enforces that the specified input and output rows for both, the old and the new node/root combinations, must be present in the trace of the hash chiplet, and that they must be exactly $8 \cdot d - 1$ rows apart, where $d$ is the depth of the node. It also ensure that the computation for the old node/root combination is immediately followed by the computation for the new node/r ...
```
nit: "The above constraint enforces... It also ensures..."

bobbinth replied:
Fixed!

bobbinth commented on docs/src/design/stack/main.md:
```diff
@@ -151,7 +145,7 @@ The above constraint can be satisfied only when either of the following holds:
 To make sure stack depth column $b_0$ is updated correctly, we need to impose the following constraints:
 
-| Condition   | Constraint       | Description |
+| Condition   | Constraint__     | Description |
```
This is a hacky way to control the width of the column. Without these underscores formulas in the column get broken into two lines.

grjte replied:
I see, thanks. I think you could do `Constraint&nbsp;&nbsp;` instead so that nothing extra shows when rendered

bobbinth commented on docs/src/design/main.md:
```diff
@@ -1,7 +1,43 @@
-  - [Bitwise Chiplet](./chiplets/bitwise.md)
-  - [Memory Chiplet](./chiplets/memory.md)
+Throughout these sections we adopt the following notations and assumptions:
+* All arithmetic operations, unless noted otherwise, are assumed to be in a prime filed with modulus $p = 2^{64} - 2^{32} + 1$.
```
Fixed!

bobbinth commented on docs/src/design/main.md:
```diff
@@ -1,7 +1,43 @@
+Throughout these sections we adopt the following notations and assumptions:
+* All arithmetic operations, unless noted otherwise, are assumed to be in a prime filed with modulus $p = 2^{64} - 2^{32} + 1$.
+* A _binary_ value means a field element which is either $0$ or $1$.
+* We use lower-case letters to refer to individual field elements (e.g., $a$), and upper-case letters to refer to groups of $4$ elements or words (e.g., $A$). To refer to individual elements within a word, we use numerical subscripts. For example, $a_0$ is the first element of word $A$, $b_3$ is the last element of word $B$, etc.
```
Fixed both!

bobbinth commented on docs/src/design/main.md:
```diff
@@ -1,7 +1,43 @@
+
+In addition to the components described previously, execution trace also contains $2$ system columns:
+
+* `clk` which is used to keep track of the current VM cycle. Values in these columns start out at $0$ and are incremented with each cycle.
```
Fixed!

bobbinth commented on docs/src/design/stack/crypto_ops.md:
```diff
@@ -0,0 +1,122 @@
+
+In the above, $op_{linhash}$ and $op_{retstate}$ are the unique [operation label](../chiplets/main.md#operation-labels) for initiating a linear hash and reading the full state of the hasher respectively. Also note that the term for $\alpha_3$ is missing from the above expressions because for Rescue Prime permutation computation the index column is expected to be set to $0$.
+
+Using the above values, we can describe constraint for the chiplet bus column as follows:
```
Fixed!

bobbinth commented on docs/src/design/stack/crypto_ops.md:
```diff
@@ -0,0 +1,122 @@
+b_{chip}' \cdot v_{input} \cdot v_{output} = b_{chip} \text{ | degree } = 3
+$$
+
+The above constrain enforces that the specified input and output rows must be present in the trace of the hash chiplet, and that they must be exactly $7$ rows apart.
```
Fixed!

bobbinth commented on docs/src/design/stack/crypto_ops.md:
```diff
@@ -0,0 +1,122 @@
+- Value of the node, 4 elements.
+- Root of the tree, 4 elements.
+
+The Merkle path itself is expected to be provided by the prover non-deterministically (via the advice provider). If the prover is not able to provide the required path, the operation fails. Otherwise, the state of the stack does not change. The diagram below illustrates this graphically.
```
Yep - agreed. I went with the dashed version everywhere.

bobbinth commented on docs/src/design/stack/crypto_ops.md:
```diff
@@ -0,0 +1,122 @@
+
+For the `MPVERIFY` operation, we define input and output values as follows:
+
+$$
+v_{input} = \alpha_0 + \alpha_1 \cdot op_{mpver} + \alpha_2 \cdot r + \alpha_3 \cdot i + \sum_{j=0}^3 \alpha_{j+8} \cdot s_{6 - j}
+$$
```
Good catch! Fixed!

bobbinth commented on docs/src/design/main.md:
```diff
@@ -1,7 +1,43 @@
+clk' - (clk + 1) = 0 \text{ | degree } = 1
+$$
+
+We frequently use the above notation for describing other AIR constraints. Specifically, the constraint is followed by its implied algebraic degree.
```
Agreed! Updated.

bobbinth commented:
I've addressed all the comments except for the ones relating to `multiset.md` and `chiplets\main.md` files. I agree with your comments there too - but I think the content there should be reworked quite a bit.

grjte commented on docs/src/design/stack/io_ops.md:
```diff
@@ -0,0 +1,193 @@
+
+To enforce the correctness of memory access, we can use the following constraint:
+
+>$$
+b_{chip}' \cdot u_{mem} = b_{chip} \text{ | degree } = 2
+$$
+
+In the above, $u_{mem}$ is the value of memory access request. Thus, to describe AIR constraint for memory operations, it is sufficient to describe how $u_{mem}$ is computed. We do this in the following sections.
```
nit: I wonder if we should use something other than `u` here, since in the Memory chiplet design docs we use `u` to refer to the previous value and `v` to refer to the new value

bobbinth replied:
I've been trying to use `u` for the "left-hand-side" of these constraints and `v` for the "right-hand-side". Basically, `u` values get divided out of the running product column and `v` values get multiplied into the running product column. I've first started using this convention for decoder constraints - but haven't really propagated it to other places yet.

This way, if we want to put together full constraints for bus columns (which include both sides), there should be no conflicts.

grjte replied:
Cool, this sounds good to me - I just wanted to make sure we're not introducing any confusion between this side of things and the way we describe things in the chiplets

grjte commented on docs/src/design/stack/io_ops.md:
```diff
@@ -0,0 +1,193 @@
+In the above, $u_{mem}$ is the value of memory access request. Thus, to describe AIR constraint for memory operations, it is sufficient to describe how $u_{mem}$ is computed. We do this in the following sections.
+
+### MLOADW
+Assume that the words with elements $v_0, v_1, v_2, v_3$ is located in memory at address $a$. The `MLOADW` operation pops an element off the stack, interprets it as a memory address, and replaces the remaining 4 elements at the top of the stack with values located at the specified address. The diagram below illustrates this graphically.
```
nit: "Assume that the word with elements..." (This comes up again below at least once)

bobbinth replied:
Fixed!

grjte commented on docs/src/design/stack/io_ops.md:
```diff
@@ -0,0 +1,193 @@
+
+The value representing state of memory before the operation (values in registers $h_0, h_1, h_2$ are set by the prover non-deterministically):
+
+$$
+v_{old} = \alpha_4 \cdot s_0' + \sum_{i=2}^3\alpha_{i+4} \cdot h_i'
+$$
```
I think there's a problem with the sum range here. Presumably everything should be the same as below, aside from the alpha indices?

bobbinth replied:
Fixed!

grjte commented on docs/src/design/stack/io_ops.md:
```diff
@@ -0,0 +1,193 @@
+
+The value representing state of memory after the operation:
+
+$$
+v_{new} = \alpha_8 \cdot s_0' + \sum_{i=0}^2\alpha_{i+8} \cdot h_i'
+$$
```
$\alpha_8$ gets used twice here. Perhaps it should be $\alpha_{i+9}$ inside the sum? (This also implies a second problem with the formula above as well as the one already mentioned)

bobbinth replied:
Fixed!

grjte commented on docs/src/design/stack/io_ops.md:
```diff
@@ -0,0 +1,193 @@
+* **No change** starting from position $1$.
+
+### MSTOREW
+Assume that the words with elements $v_0, v_1, v_2, v_3$ is located in memory at address $a$. The `MSTOREW` operation pops an element off the stack, interprets it as a memory address, and writes the remaining $4$ elements at the top of the stack into memory at the specified address. The stored elements are not removed from the stack. The diagram below illustrates this graphically.
+
+![mstorew](../../assets/design/stack/io_ops/MSTOREW.png)
+
+After the operation the contents of memory at address $a$ would be set to $u_0, u_1, u_2, u_3$.
```
nit: This is backwards from how we specify old/new values in the Memory chiplet design docs, which I think is a bit confusing. (Over there we use `u` for old and `v` for new.) I'd prefer to be consistent with that or to change that to be consistent with this.

bobbinth replied:
Agreed! I've updated them here to match how it is in memory chiplet description.

grjte commented on docs/src/design/stack/io_ops.md:
```diff
@@ -0,0 +1,193 @@
+
+The value representing state of memory after the operation:
+
+$$
+v_{new} = \alpha_4 \cdot s_0' + \sum_{i=1}^3\alpha_{i+8} \cdot h_i'
+$$
```
I think this should be $\alpha_8$ instead of $\alpha_4$

bobbinth replied:
Fixed!

grjte reviewed (changes requested):
I've still got a couple more of the operations files to go through, so this is another partial review.

bobbinth commented on docs/src/design/stack/field_ops.md:
```diff
@@ -0,0 +1,178 @@
+The effect on the rest of the stack is:
+* **Left shift** starting from position $2$.
+
+## OR
+Assume $a$ and $b$ are the elements at the top of the stack. The `OR` operation computes $c \leftarrow (a \lor b)$ The diagram below illustrates this graphically.
+
+![or](../../assets/design/stack/field_operations/OR.png)
+
+Stack transition for this operation must satisfy the following constraints:
+
+>$$
+s_i^2 - s_i = 0 \text{ for } i \in \{0, 1\} \text{ | degree } = 2
+$$
+
+>$$
+s_{0}' - (s_{1} + s_{0} - s_{1} \cdot s_{0}) = 0 \text{ | degree } = 2
+$$
+
+The first two constraints ensure that the value in $s_0$ and $s_1$ are binary, and the third constraint ensures the correctness of the boolean `OR` operation.
+
+The effect on the rest of the stack is:
+* **Left shift** starting from position $2$.
```
We don't have a dedicated instruction for binary `XOR`, instead it is emulated in the assembly using `AND` and `OR` operations. (this is different for u32 operations where we do have a dedicated `XOR`).

grjte replied:
Ah yes, sorry about that!

bobbinth commented on docs/src/design/stack/field_ops.md:
```diff
@@ -0,0 +1,178 @@
+The effect on the rest of the stack is:
+* **No change** starting from position $1$.
+
+## NOT
+Assume $a$ is the element at the top of the stack. The `NOT` operation computes $b \leftarrow (\lnot a)$. The diagram below illustrates this graphically.
```
Agreed! Updated.

bobbinth commented:
I've addressed the above comments and also updated the AIR constraint portion of the memory design note (to make it consistent with IO operation design descriptions).

grjte commented on docs/src/design/stack/u32_ops.md:
```diff
@@ -0,0 +1,294 @@
+
+To facilitate this operation, we will need to make a request to the chiplet bus $b_{chip}$ by dividing its current value by the value representing bitwise operation request. This can be enforced with the following constraint:
+
+>$$
+b_{chip}' \cdot \left(\alpha_0 + \alpha_1 \cdot op_{u32and} + \alpha_2 \cdot s_0 + \alpha_3 \cdot s_1 +  \alpha_4 \cdot s_0'  \right) = b_{chip} \text{ | degree } = 2
+$$
```
I believe we've previously described this as shown below, which would have the values of a and b swapped from what's described / pictured here. (I just want to make sure we're being consistent between here and the Bitwise chiplet). Same comment for `U32OR` and `U32XOR`.

$$
b_{chip}' \cdot \left(\alpha_0 + \alpha_1 \cdot op_{u32and} + \alpha_2 \cdot a + \alpha_3 \cdot b +  \alpha_4 \cdot c  \right) = b_{chip} \text{ | degree } = 2
$$

bobbinth replied:
Yep - I've updated the diagrams to be consistent with the above formula.

grjte commented on docs/src/design/chiplets/memory.md:
```diff
@@ -197,110 +195,54 @@ The amortized cost of reading or writing a single value is between $4$ and $5$ t
 $$
 
-where $i \in \{0, 1, 2, 3\}$.
+On the stack side, for every memory access request, a corresponding value is divided out of the $b_{chip}$ column. Specifics of how this is done are described [here](../stack/io_ops.md#memory-access-operations).
```
This new constraint section description is an improvement overall, but I think this last bit is confusing, since we haven't provided the "corresponding value" here and specified that it needs to be multiplied into $b_{chip}$ on this side. We've only provided the constraints required for keeping the memory chiplet internally consistent. I think it's important to include information about the lookup here as well, with the calculation of the memory row value. For people trying to understand how these things are connected, I think seeing the formula and a brief reference to it in both places with links to the corresponding section will be helpful (at least, it would have been for me)

bobbinth replied:
Ah - I've actually written a whole sub-section about this - but must have been over eager with deleting things and it didn't get committed. I've restored it now.

grjte reviewed (approved):
Here's the rest of my review. Although some of the comments are things that do need to be addressed (e.g. incorrect constraint descriptions or images), I think they're all pretty straightforward, so I don't think a follow-up review will be necessary unless you want input on any of the changes (in which case I'm happy to). Therefore, I'm marking it approved, assuming these comments will be addressed

bobbinth commented on docs/src/design/stack/stack_ops.md:
```diff
@@ -0,0 +1,225 @@
+## SWAPDW
+The `SWAPDW` operation swaps stack elements $\{0, ..., 7\}$ with elements $\{8, ..., 15\}$. The diagram below illustrates this graphically.
+
+![swapdw](../../assets/design/stack/stack_ops/SWAPDW.png)
```
Good catch! Fixed!

bobbinth commented on docs/src/design/stack/stack_ops.md:
```diff
@@ -0,0 +1,225 @@
+* **Left shift** starting from position $3$.
+
+## CSWAPW
+The `CSWAPW` operation pops an element off the stack and if the element is $1$, swaps elements $0, 1, 2, 3$ with elements $4, 5, 6, 7$. If the popped element is $0$, the stack remains unchanged. The diagram below illustrates this graphically.
```
Agreed! Updated.

bobbinth commented on docs/src/design/stack/u32_ops.md:
```diff
@@ -0,0 +1,294 @@
+### Checking element validity
+Another primitive which most operations described below rely is checking whether four 16-bit values form a valid field element. Assume $t_0$, $t_1$, $t_2$, and $t_3$ are known to be 16-bit values, and we want to verify that $2^{48} \cdot t_3 + 2^{32} \cdot t_2 + 2^{16} \cdot t_1 + t_0$ is a valid field element. 
+
+For simplicity, let's denote:
+
+$$
+v_{hi} = 2^{16} \cdot t_3 + t_2 \\
+v_{lo} = 2^{16} \cdot t_1 + t_0
+$$
```
Agreed! Updated.

bobbinth commented on docs/src/design/stack/op_constraints.md:
```diff
@@ -0,0 +1,250 @@
+Using the operation flags defined above, we can compute several composite flags which are used by various constraints in the VM.
+
+### Shift right flag
+The right-shift flag indicates that a given operation shifts the stack to the right. This flag is computed as follows:
+
+$$
+f_{shr} = (1 - b_6) \cdot b_5 \cdot b_4 + f_{u32split} + f_{push}
+$$
```
Updated!
